scripts/20210517_validation-object/script.js

```diff
@@ -0,0 +1,283 @@
+// Expect 0 records
+
+db.getCollection('forms').aggregate([
+  { $unwind: '$form_fields' },
+  {
+    $match: {
+      'form_fields.fieldType': { $in: ['textfield', 'textarea', 'number'] },
+    },
+  },
+  {
+    $project: {
+      form_fields: 1,
+      selectedValidation: '$form_fields.ValidationOptions.selectedValidation',
+      customVal: '$form_fields.ValidationOptions.customVal',
+      customMin: '$form_fields.ValidationOptions.customMin',
+      customMax: '$form_fields.ValidationOptions.customMax',
+    },
+  },
+  {
+    $match: {
+      $and: [
+        {
+          selectedValidation: 'Minimum',
+        },
+        {
+          $expr: { $ne: ['$customMin', '$customVal'] },
+        },
+      ],
+    },
+  },
+])
+
+// b. customVal === customMax when selectedValidation === 'Maximum'
+// Expect 0 records
+
+db.getCollection('forms').aggregate([
+  { $unwind: '$form_fields' },
+  {
+    $match: {
+      'form_fields.fieldType': { $in: ['textfield', 'textarea', 'number'] },
+    },
+  },
+  {
+    $project: {
+      form_fields: 1,
+      selectedValidation: '$form_fields.ValidationOptions.selectedValidation',
+      customVal: '$form_fields.ValidationOptions.customVal',
+      customMin: '$form_fields.ValidationOptions.customMin',
+      customMax: '$form_fields.ValidationOptions.customMax',
+    },
+  },
+  {
+    $match: {
+      $and: [
+        {
+          selectedValidation: 'Maximum',
+        },
+        {
+          $expr: { $ne: ['$customMax', '$customVal'] },
+        },
+      ],
+    },
+  },
+])
+
+// c. customVal === customMin === customMax when selectedValidation === 'Exact'
+// Expect 0 records
+
+db.getCollection('forms').aggregate([
+  { $unwind: '$form_fields' },
+  {
+    $match: {
+      'form_fields.fieldType': { $in: ['textfield', 'textarea', 'number'] },
+    },
+  },
+  {
+    $project: {
+      form_fields: 1,
+      selectedValidation: '$form_fields.ValidationOptions.selectedValidation',
+      customVal: '$form_fields.ValidationOptions.customVal',
+      customMin: '$form_fields.ValidationOptions.customMin',
+      customMax: '$form_fields.ValidationOptions.customMax',
+    },
+  },
+  {
+    $match: {
+      $and: [
+        {
+          selectedValidation: 'Exact',
+        },
+        {
+          $or: [
+            {
+              $expr: { $ne: ['$customMax', '$customVal'] },
+            },
+            {
+              $expr: { $ne: ['$customMin', '$customVal'] },
+            },
+          ],
+        },
+      ],
+    },
+  },
+])
+
+// d. selectedValidation === 'Range' should not exist
+// Expect 0 records
+
+db.getCollection('forms').aggregate([
+  { $unwind: '$form_fields' },
+  {
+    $match: {
+      'form_fields.fieldType': { $in: ['textfield', 'textarea', 'number'] },
+    },
+  },
+  {
+    $project: {
+      form_fields: 1,
+      selectedValidation: '$form_fields.ValidationOptions.selectedValidation',
+    },
+  },
+  {
+    $match: {
+      selectedValidation: 'Range',
+    },
+  },
```

Comment:
I have a feeling we can skip all the previous steps and just do `{ $match: { 'form_fields.ValidationOptions.selectedValidation': 'Range' } }`, since we shouldn't have such a key regardless of fieldType

Reply:
that is true with an `elemMatch`... but I guess I will keep the current implementation because it's more 'precise' :)

Reply:
okay sure! I don't think you need elemMatch here either though